### PR TITLE
AdEx v5: Issue #384 Unified precision - primitives::UnifiedNum

### PR DESCRIPTION
**Pull request by elpiel**
Changelog:
- Add `primitives::UnifiedNum` for #384 with precision of 8

### CODE REVIEW

**Comment by samparsky** [comment]
what's the reason for having a `Unified precision`?

**Comment by elpiel** [comment]
### Reason
> what's the reason for having a `Unified precision`?

The way I understand it is that this will help us to have a unified representation of values that is good enough for all applications. We won't need more precision than this, it will ease the calculations and we can convert to the on-chain values when we create the MerkleTree we can use the full precision of the stablecoin.

Maybe @Ivshti could give a fuller answer here?

### Why 8 in particular

From AIP#61:
> validator: unified precision: use a unified number of decimals for all numbers (8), and only convert to the proper stablecoin decimals when generating the balance tree; this will ensure that we can handle precision even with USDT (6 decimals); normal int numbers can be used in DBs (easier, can aggregate stuff values faster)


If the CPM price is 0.01 then with a precision of 8 and a fee of 1% we will get a value of 0.00000001 in Unified Precision

**Comment by Ivshti** [comment]
The main reason behind using unified precision is being able to use easy-to-work with and fast types such as u64. 

Furthermore, it keeps numbers in the database consistent if we switch between stablecoins with different decimal precision, and we don't need to account for that when calculating stats/aggregate data.

But most importantly, using unified precision allows us to use native number types in databases (mongo, postres), allowing for much easier updates and aggregations, opening the door towards more analytics features.

There's no real reason for supporting more than 8 decimals.

**EDIT:** sorry, closed by accident

File: primitives/src/unified_num.rs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        // 321.00000777 - should floor to 321.000007 (precision 6)
+        let usdt_unified = UnifiedNum::from(32_100_000_777_u64);
+        let usdt_expected = BigNum::from(321_000_007_u64);
+        assert_eq!(
+            usdt_expected,
+            usdt_unified.to_precision(usdt_precision),
+            "It should floor the result of USDT"
+        );
+
+        // 321.00000999
```
Comment (samparsky) [inline primitives/src/unified_num.rs]: should be 999 or 7777?

**Review by samparsky — APPROVED** [review]

File: primitives/src/unified_num.rs (hunk 1 above)
Comment (elpiel) [reply primitives/src/unified_num.rs]: good catch!

**Pull request merged**